src/mlpack/methods/cf/cf_main.cpp

```diff
@@ -138,21 +173,135 @@ void ComputeRecommendations(CFModel* cf,
     if (users.n_rows > 1)
       users = users.t();
     if (users.n_rows > 1)
-      Log::Fatal << "List of query users must be one-dimensional!" << std::endl;
-
-    Log::Info << "Generating recommendations for " << users.n_elem << " users."
-        << endl;
-    cf->GetRecommendations(numRecs, recommendations, users.row(0).t());
+      Log::Fatal << "List of query users must be one-dimensional!"
+                 << std::endl;
+
+    Log::Info << "Generating recommendations for "
+              << users.n_elem << " users."
+              << endl;
+
+    //  All possible alternatives for Recommendations
+    if (neighborSearchAlgorithm == "Cosine")
```

Comment:
There's actually a nicer and more flexible way to do this that will help expand this in the future.  I think that the idea can also be found in `kmeans_main.cpp`.  Basically it looks like this...

```
void GetRecommendations(CF* c, const size_t numRecs, arma::Mat<size_t>& recommendations)
{
  const string& neighborSearchAlgorithm = CLI::GetParam<string>("neighbor_search");

  if (neighborSearxhAlgorithm == "Cosine")
    GetRecommendations<CosineSearch>(c, numRecs, recommendations);
  else if (neighborSearchAlgorithm == "Euclidean")
    GetRecommendations<EuclideanSearch>(c, numRecs, recommendations);
  ...
}

template<typename NeighborSearchType>
void GetRecommendations(CF* c, const size_t numRecs, arma::Mat<size_t>& recommendations)
{
  const string& interpolationAlgorithm = CLI::GetParam<string>("interpolation_algorithm");

  if (interpolationAlgorithm == "Average")
    GetRecommendations<NeighborSearchType, AverageInterpolation>(c, numRecs, recommendations);
  else if ...
}

template<typename NeighborSearchType, typename InterpolationType>
void GetRecommendations(CF* c, const size_t numRecs, arma::Mat<size_t>& recommendations)
{
  c->GetRecommendations<NeighborSearchType, InterpolationType>(numRecs, recommendations);
}
```

You could do the same for `Predict()`.  Sorry for any errors in the idea code---I wrote this on a phone so maybe it is not perfect. :)

Reply:
Thanks for taking the time and writing this part, I looked at it and also `kmeans_main.cpp`. I've updated the PR as per it.